Improvement/3088 overview metrics

alexis-ld opened the pull request:
**Component**: ui, dashboard

<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->

**Context**: 
We want to add a metrics panel in the dashboard page. More infos on #3088 

**Summary**:


https://user-images.githubusercontent.com/3480526/115054259-54f2e700-9ee0-11eb-9651-48b14986a085.mov



---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #3088 
Closes: #3345 
<!-- If you want to refer to an issue while not closing it, use:

See: #ISSUE_NUMBER

-->

bert-e commented:
# Hello alexis-ld,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nicolast

* @chengyanjin

* @alexandre-allard-scality

* @thomasdanan

* @gdemonet

* @teddyandrieux

* @lucieleonard

* @jbwatenbergscality

ChengYanJin commented on ui/src/components/DashboardMetrics.js:
```diff
@@ -0,0 +1,234 @@
+
+const DashboardMetrics = () => {
+  // Making sure nodes list is loaded
+  useRefreshEffect(refreshNodesAction, stopRefreshNodesAction);
+  const nodes = useTypedSelector((state) => state.app.nodes.list);
+  const theme = useTypedSelector((state) => state.config.theme);
```
We would eventually get rid of redux.  So I suggest using `react-query` to achieve it.

ChengYanJin commented on ui/src/components/DashboardMetrics.js:
```diff
@@ -0,0 +1,234 @@
+    axis: {
+      // Refer to all the available time format: https://github.com/d3/d3-time-format#locale_format
+      format:
+        metricsTimeSpan === (LAST_ONE_HOUR || LAST_TWENTY_FOUR_HOURS)
+          ? '%H:%M'
+          : '%m/%d',
```
From the screenshot, I see the x-axis is always the same date, not sure what's the timespan though. Please add the timespan selector with three choices.

alexis-ld replied:
For now it's static on LAST_TWENTY_FOUR_HOURS. The design does not show a timespan selector so I figured it would be static.

@ChengYanJin @Cuervino there is not much space between the panel title and the Advanced metrics button on the design, where should I put this ?

ChengYanJin commented on ui/src/containers/DashboardPage.js:
```diff
@@ -26,6 +28,10 @@ const DashboardGrid = styled.div`
   div {
     background-color: ${(props) => props.theme.brand.primary};
     color: ${(props) => props.theme.brand.textPrimary};
+    padding: ${padding.smaller};
+    .sc-vegachart svg {
+      background-color: inherit !important;
+    }
   }
```
Better go inside the `<DashboardMetrics/>`component. Since it's the detail about this component.

alexis-ld replied:
@ChengYanJin I put it there because we will have charts in the Network panel and we will need the same background on those as well. :) Should I still move it and then copy the code in both components ?

ChengYanJin commented on ui/src/components/DashboardMetrics.js:
```diff
@@ -0,0 +1,234 @@
+            lineConfig={lineConfig}
+            tooltip={true}
+            tooltipConfig={perNodeTooltip}
+            tooltipTheme={'dark'}
```
We added a custom tooltip for vega-tooltip. 
Once we release a new version of `core-ui`, you can benefit from it.
```tooltipTheme={'custom'}```

bert-e commented:
# Branches have diverged



This pull request's source branch `improvement/3088-overview-metrics` has diverged from
`development/2.9` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/2.9` into `improvement/3088-overview-metrics`
* Rebase `improvement/3088-overview-metrics` onto `origin/development/2.9`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nicolast

* @chengyanjin

* @alexandre-allard-scality

* @thomasdanan

* @gdemonet

* @teddyandrieux

* @lucieleonard

* @jbwatenbergscality

alexis-ld commented:
@ChengYanJin Switching to ready for review PR removing Draft.
Added Throughput chart, panel title and Advanced metrics button.
As we discussed I also removed the timespan selector since it will be handled at the top level of the dashboard;

Also it seems the overflowing elements at the bottom of the page are not visible. But as we discussed it does not matter since the layout will soon be updated.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/3088-overview-metrics` into
`development/2.9`
* [w/2.10/improvement/3088-overview-metrics](https://github.com/scality/metalk8s/compare/development/2.10...w/2.10/improvement/3088-overview-metrics) will be merged into `development/2.10` 



The following branches will **NOT** be impacted:

* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @gdemonet

* @chengyanjin

* @teddyandrieux

* @jbwatenbergscality

* @alexandre-allard-scality

* @nicolast

* @thomasdanan

* @lucieleonard

ChengYanJin commented on ui/src/services/utils.js:
```diff
@@ -419,18 +419,21 @@ export const useTableSortURLSync = (sorted, desc, data, defaultSortKey) => {
-    if (graphsContainerWidth) setGraphWidth(graphsContainerWidth / 2 - 50);
-  }, [graphsContainerWidth]);
+    if (graphsContainerWidth) {
+      if (columns === 1) setGraphWidth(graphsContainerWidth - 100);
```
I think the width of the charts should be bigger to fully use the space on the right.

ChengYanJin commented on ui/src/components/DashboardMetrics.js:
```diff
@@ -0,0 +1,443 @@
+  justify-content: space-between;
+`;
+
+// import styled from 'styled-components';
```
can be removed

ChengYanJin commented on ui/src/components/DashboardMetrics.js:
```diff
@@ -0,0 +1,443 @@
+    '#245A83',
+    '#808080',
+    '#A04EC9',
+    '#C6B38A',
```
I would suggest we create a global constant for those colors and name them maybe `linePrimary`, `lineSecondary` since we don't have those colors in the color palette yet.

ChengYanJin commented on ui/src/components/DashboardMetrics.js:
```diff
@@ -0,0 +1,443 @@
+        )}
+        <LineChart
+          id={'dashboard_memory_id'}
+          data={memoryDataQuery.isSuccess ? memoryDataQuery.data : noData}
```
are you sure we need this given
```
  useEffect(() => {
    if (!nodesQuery.isLoading && nodesQuery.isSuccess) {
      setNodes(nodesQuery.data);
    }
  }, [nodesQuery]);
  ```
anyway, we should display the loader if the data is an empty array, no?

ChengYanJin commented on ui/src/components/DashboardMetrics.js:
```diff
@@ -0,0 +1,443 @@
+          data={memoryDataQuery.isSuccess ? memoryDataQuery.data : noData}
+          xAxis={xAxis}
+          yAxis={yAxisUsage}
+          color={memoryDataQuery.isSuccess ? perNodeColor : null}
```
not sure we need this either. what's the reason?

ChengYanJin commented on ui/src/services/utils.js:
```diff
@@ -419,18 +419,21 @@ export const useTableSortURLSync = (sorted, desc, data, defaultSortKey) => {
-  }, [graphsContainerWidth]);
+    if (graphsContainerWidth) {
+      if (columns === 1) setGraphWidth(graphsContainerWidth - 100);
+      else setGraphWidth(graphsContainerWidth / columns - 90);
```
and we suggest not to change the layout for the node metrics and volume metrics page.

ChengYanJin reviewed (changes requested):
As well as some issues I already sent you:
- Fix the small responsiveness issue in order to see the legends.
- Errors pop up in the console
  1. ERROR Error: invalid bounds
  2. vega-scenegraph.module.js:4680 Error: <svg> attribute viewBox: Expected number, "0 0 1.79769313486231…".


Please investigate the error, and let me know if you need any help.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @gdemonet

* @chengyanjin

* @teddyandrieux

* @jbwatenbergscality

* @alexandre-allard-scality

* @nicolast

* @thomasdanan

* @lucieleonard





The following reviewers are expecting changes from the author, or must review again:

* @chengyanjin

bert-e commented:
# History mismatch



Merge commit #cde33d8cfc4a753ec6e80d3ba156935956ac5a48 on the integration branch
`w/2.10/improvement/3088-overview-metrics` is merging a branch which is neither the current
branch `improvement/3088-overview-metrics` nor the development branch
`development/2.10`.

It is likely due to a rebase of the branch `improvement/3088-overview-metrics` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

ChengYanJin commented on ui/src/components/DashboardMetrics.js:
```diff
@@ -0,0 +1,460 @@
+            queryNodeCPUMetrics(node.internalIP, metricsTimeSpan),
+          ),
+        ).then((result) => formatNodesPromRangeForChart(result)),
+      [nodes, metricsTimeSpan, formatNodesPromRangeForChart],
```
what's the reason we put `formatNodesPromRangeForChart` as useCallback dependency?

alexis-ld replied:
@ChengYanJin to avoid the warning: 
` React Hook useCallback has a missing dependency: 'formatNodesPromRangeForChart'. Either include it or remove the dependency array  react-hooks/exhaustive-deps`

React wants to know it needs to update the function if formatNodesPromRangeForChart changes (it won't happen) but it considers it's a good practice to include it.

ChengYanJin commented on ui/src/components/DashboardMetrics.js:
```diff
@@ -0,0 +1,460 @@
+
+  const lineConfig = { strokeWidth: 1.5 };
+
+  const formatNodesPromRangeForChart = useCallback(
```
I suggest we add unit test for `formatNodesPromRangeForChart` and `formatNodesThroughputPromRangeForChart`

ChengYanJin commented on ui/src/components/DashboardMetrics.js:
```diff
@@ -0,0 +1,460 @@
+
+  const perNodeTooltip = getTooltipConfig(
+    nodes.map((node) => ({
+      field: node.name,
```
The value in the tooltips is NaN :(

ChengYanJin commented on ui/src/hooks.js:
```diff
@@ -10,3 +14,35 @@ export const useTypedSelector: <TSelected>(
+export const useNodes = (): Object[] => {
+  const [nodes, setNodes] = useState([]);
+
+  const nodesQuery = useQuery(['nodesNames'], () =>
```
```suggestion
  const nodesQuery = useQuery('nodesNames', () =>
```

ChengYanJin reviewed (changes requested):
1. I would not expect to reload the graph when I switch back to the dashboard page from the Node page for example. Some optimation needs to be done around `react-query`. 

2. We should have graph refresh every 1 min(REFRESH_METRCIS_GRAPH). Base on the current implementation I think we miss this functionality.

3. I still can reproduce the console error `ERROR Error: invalid bounds` by the following path:
Go to dashboard page => go to node page => go back to dashboard page.

alexis-ld commented:
/approve

bert-e commented:
# History mismatch



Merge commit #cde33d8cfc4a753ec6e80d3ba156935956ac5a48 on the integration branch
`w/2.10/improvement/3088-overview-metrics` is merging a branch which is neither the current
branch `improvement/3088-overview-metrics` nor the development branch
`development/2.10`.

It is likely due to a rebase of the branch `improvement/3088-overview-metrics` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

alexis-ld commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/3088-overview-metrics` into
`development/2.9`
* [w/2.10/improvement/3088-overview-metrics](https://github.com/scality/metalk8s/compare/development/2.10...w/2.10/improvement/3088-overview-metrics) will be merged into `development/2.10` 



The following branches will **NOT** be impacted:

* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @gdemonet

* @chengyanjin

* @teddyandrieux

* @jbwatenbergscality

* @alexandre-allard-scality

* @nicolast

* @thomasdanan

* @lucieleonard





The following reviewers are expecting changes from the author, or must review again:

* @chengyanjin







*The following options are set:* **approve**

JBWatenbergScality commented on ui/src/components/DashboardMetrics.js:
```diff
@@ -0,0 +1,456 @@
+} from '../services/utils';
+import { expressionFunction } from 'vega';
+
+// Custom formatter to display negative value as absolut value in throughput chart
```
```suggestion
// Custom formatter to display negative value as an absolute value in throughput chart
```

JBWatenbergScality commented on ui/src/components/DashboardMetrics.js:
```diff
@@ -0,0 +1,456 @@
+  justify-content: space-between;
+`;
+
+export const formatNodesPromRangeForChart = (result, nodes) => {
```
We would like to enforce the typing in order to prevent some predictable errors. Could you please add it for the arguments and the return type ?

JBWatenbergScality commented on ui/src/components/DashboardMetrics.js:
```diff
@@ -0,0 +1,456 @@
+  return reduced;
+};
+
+export const formatNodesThroughputPromRangeForChart = (result, nodes) => {
```
We would like to enforce the typing in order to prevent some predictable errors. Could you please add it for the arguments and the return type ?

JBWatenbergScality commented on ui/src/components/DashboardMetrics.js:
```diff
@@ -0,0 +1,456 @@
+  const queryTimeSpan = query.get('from');
+
+  const [metricsTimeSpan, setMetricsTimeSpan] = useState(
+    LAST_TWENTY_FOUR_HOURS,
```
Later on we want the time span to be the same one across the application if the user change it. I believe we should introduce a context and a custom hook to retrieve the selected / set the chart time span so that component can use the same one globally available.

JBWatenbergScality replied:
Once the time span mutualised in a context we can also move the query parma from behavior to that global context and avoid duplicating this behavior in all the components using charts

JBWatenbergScality commented on ui/src/components/DashboardMetrics.js:
```diff
@@ -0,0 +1,456 @@
+    darken(0.2, lineColor4),
+  ];
+
+  const getThroughputColorRange = (size) => {
```
Please add some typing for the arguments and return type

JBWatenbergScality commented on ui/src/components/DashboardMetrics.js:
```diff
@@ -0,0 +1,456 @@
+      // Refer to all the available time format: https://github.com/d3/d3-time-format#locale_format
+      format:
+        metricsTimeSpan === LAST_TWENTY_FOUR_HOURS ||
+        metricsTimeSpan === LAST_ONE_HOUR
```
I suggest to switch this implementation to an interval base time selection as we may want to give more control to the user later on to select a specific time period

JBWatenbergScality commented on ui/src/components/DashboardMetrics.js:
```diff
@@ -0,0 +1,456 @@
+  return null;
+};
+
+const DashboardMetrics = () => {
```
In order to simplify maintainability of this component I would suggest to extract all the graphs related functions to a specific component

JBWatenbergScality commented on ui/src/components/DashboardMetrics.js:
```diff
@@ -0,0 +1,456 @@
+        Promise.all(
+          nodes.map((node) =>
+            queryNodeCPUMetrics(node.internalIP, metricsTimeSpan),
+          ),
```
Those metrics are also loaded in the node page individually so I think it would make sense to use `useQueries` here (https://react-query.tanstack.com/reference/useQueries) in order to avoid re fetching the same data when switching to the node page and mutualise caching

JBWatenbergScality commented on ui/src/components/DashboardMetrics.js:
```diff
@@ -0,0 +1,456 @@
+        Promise.all(
+          nodes.map((node) =>
+            queryNodeMemoryMetrics(node.internalIP, metricsTimeSpan),
+          ),
```
Similarly to the CPU metrics it would makes sense to use `useQueries` here

JBWatenbergScality commented on ui/src/components/DashboardMetrics.js:
```diff
@@ -0,0 +1,456 @@
+        Promise.all(
+          nodes.map((node) =>
+            queryNodeLoadMetrics(node.internalIP, metricsTimeSpan),
+          ),
```
Similarly to the CPU metrics it would makes sense to use useQueries here

JBWatenbergScality commented on ui/src/components/DashboardMetrics.js:
```diff
@@ -0,0 +1,456 @@
+      );
+      if (formatted && formatted.value) setMetricsTimeSpan(formatted.value);
+    }
+  }, [queryTimeSpan]);
```
To move to the time span context when applying the previous comment

JBWatenbergScality commented on ui/src/containers/DashboardPage.js:
```diff
@@ -42,20 +53,99 @@ const DashboardGrid = styled.div`
 
 const DashboardPage = (props: {}) => {
+  const history = useHistory();
+  const query = useQuery();
```
nit: we might want to rename this hook in order to avoid confusion with react-query when reviewing the code

JBWatenbergScality commented on ui/src/containers/DashboardPage.js:
```diff
@@ -42,20 +53,99 @@ const DashboardGrid = styled.div`
+  }, [queryTimeSpan]);
+
+  // Write the selected timespan in URL
+  const writeUrlTimeSpan = (timespan) => {
```
Please add some typing here

JBWatenbergScality commented on ui/src/hooks.js:
```diff
@@ -10,3 +14,43 @@ export const useTypedSelector: <TSelected>(
+/**
+ * It retrieves the nodes data through react-queries
+ */
+export const useNodes = (): Object[] => {
```
We should avoid to use Object type as this is too generic.

JBWatenbergScality commented on ui/src/services/prometheus/api.js:
```diff
@@ -69,7 +69,7 @@ export function queryPrometheusRange(
   step: number, // step (time in seconds between 2 metrics)
   query: string, // query (the actual PromQL expression)
-) {
+): function {
```
function type is too generic. In practice here it is returning a Promise< PrometheusQueryResult >

JBWatenbergScality commented on ui/src/services/utils.js:
```diff
@@ -421,18 +421,21 @@ export const useTableSortURLSync = (sorted, desc, data, defaultSortKey) => {
-  const graphsContainerWidth = document.getElementById(container_id)
-    ?.offsetWidth;
-  const [graphWidth, setGraphWidth] = useState(0);
+export const useDynamicChartSize = (container_id, columns = 2, rows = 3) => {
```
Please add some typing here

bert-e commented:
# Conflict




A conflict has been raised during the update of
integration branch `w/2.10/improvement/3088-overview-metrics` with contents from `improvement/3088-overview-metrics`
and `development/2.10`.

Please resolve the conflict on **the integration branch** (`w/2.10/improvement/3088-overview-metrics`).


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout w/2.10/improvement/3088-overview-metrics
 $ git pull  # or "git reset --hard origin/w/2.10/improvement/3088-overview-metrics"
 $ git merge origin/development/2.10
 $ # <intense conflict resolution>
 $ git commit
 $ git merge origin/improvement/3088-overview-metrics
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.10/improvement/3088-overview-metrics
```





*The following options are set:* **approve**

bert-e commented:
# History mismatch



Merge commit #f129c9dc088e38f52c1708a3a1ca88c52c88de23 on the integration branch
`w/2.10/improvement/3088-overview-metrics` is merging a branch which is neither the current
branch `improvement/3088-overview-metrics` nor the development branch
`development/2.10`.

It is likely due to a rebase of the branch `improvement/3088-overview-metrics` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @gdemonet

* @chengyanjin

* @teddyandrieux

* @jbwatenbergscality

* @alexandre-allard-scality

* @nicolast

* @thomasdanan

* @lucieleonard





The following reviewers are expecting changes from the author, or must review again:

* @chengyanjin







*The following options are set:* **approve**

ChengYanJin commented on ui/src/components/DashboardMetrics.test.js:
```diff
@@ -0,0 +1,262 @@
+import {
```
Thanks for adding the test!
However, for me `formatNodesPromRangeForChart` is more a util function. 
`DashboardMetrics.test.js` would make me think it's the unit test for the entire component.
Hence, I would suggest moving the test to `graphUtil.test.js`

ChengYanJin commented on ui/src/components/DashboardChartCpuUsage.js:
```diff
@@ -0,0 +1,68 @@
+  lineConfig: Object,
+  perNodeTooltip: Object,
+  reactQueryOptions: UseQueryOptions,
+};
```
Do we have exactly the same props for all the charts in the dashboard page? Do you think we can share it?

ChengYanJin commented on ui/src/components/DashboardMetrics.js:
```diff
@@ -0,0 +1,260 @@
+  return reduced;
+};
+
+export const formatNodesThroughputPromRangeForChart = (
```
I would suggest extracting this util function to a dedicated file - `graphUtil.js` 
Same as `formatNodesPromRangeForChart`

ChengYanJin commented on ui/src/components/DashboardMetrics.js:
```diff
@@ -0,0 +1,260 @@
+    reactQueryOptions: {
+      refetchOnMount: false,
+      refetchOnWindowFocus: false,
+      refetchInterval: 60000,
```
I suggest we use `REFRESH_METRCIS_GRAPH` from constant.js

ChengYanJin commented on ui/src/components/DashboardMetrics.test.js:
```diff
@@ -0,0 +1,262 @@
+
+const testFailedData = [
+  {
+    status: 'fail',
```
```suggestion
    status: 'error',
```

ChengYanJin commented on ui/src/types.js:
```diff
@@ -1,12 +1,12 @@
-  name: V1ObjectMeta.name | undefined,
-};
+  internalIP: V1NodeAddress.address,
+  name: V1ObjectMeta.name,
```
Not sure this is correct.
If you want to typing `instanceIP`, you can directly type `string`

ChengYanJin commented on ui/src/hooks.js:
```diff
@@ -60,3 +61,55 @@ export const useNodes = (): Nodes => {
+  );
+};
+
+export const useMetricsTimeSpan = (): [
```
I am wondering if it's possible to add some unit tests for this hook?

ChengYanJin reviewed (changes requested):
Some small changes request, and please make sure to clean the commit in the end.
Also, in my opinion, we should avoid this reload as much as possible since we already fetch the metrics.
https://user-images.githubusercontent.com/18453133/120627386-e8b64d80-c463-11eb-8ee3-64fe9b6554b8.mov

For this point, one solution I see is not to render the graph until the graphWidth!==0.

ChengYanJin commented:
Last point: 
Please double with design, if the paddings of the `Advanced button` and the `graph title` are correct.

ChengYanJin commented on ui/src/services/graphUtils.test.js:
```diff
@@ -0,0 +1,262 @@
+
+const testFailedThroughputData = [
+  {
+    status: 'fail',
```
```suggestion
    status: 'error',
```

alexis-ld committed:
ui, metrics: Add fetch method for dashboard metrics

Refs: #3088

alexis-ld committed:
ui: Rename useQuery to useURLQuery + fix chart sizing hook

Refs: #3088

alexis-ld committed:
ui, monitoring: Fix typo in constant + timespan constants

Refs: #3088

ChengYanJin commented on ui/src/services/prometheus/fetchMetrics.js:
```diff
@@ -30,3 +33,139 @@ export function queryNodeFSSize(
+};
+
+export function queryNodeCPUMetrics(
+  instanceIP: $PropertyType<NodeAddress, 'internalIP'>,
```
```suggestion
  instanceIP: string,
```

ChengYanJin commented on ui/src/services/prometheus/fetchMetrics.js:
```diff
@@ -30,3 +33,139 @@ export function queryNodeFSSize(
+}
+
+export function queryNodeMemoryMetrics(
+  instanceIP: $PropertyType<NodeAddress, 'internalIP'>,
```
```suggestion
  instanceIP: string,
```

ChengYanJin commented on ui/src/hooks.js:
```diff
@@ -10,3 +22,97 @@ export const useTypedSelector: <TSelected>(
+  return nodes;
+};
+
+export const useNodeAddressesSelector = (nodes: Nodes): NodesAddresses => {
```
```suggestion
export const useNodeAddressesSelector = (nodes: Nodes): Array<{internalIP: string, name: string}>=> {
```
IMO, I would be easier to read if we directly type here.

ChengYanJin commented on ui/src/types.js:
```diff
@@ -1,3 +1,13 @@
 //@flow
+import { V1Node } from '@kubernetes/client-node';
```
I think we can directly use `V1NodeList` to type the return of `useNodes()`
import { V1NodeList } from '@kubernetes/client-node';

ChengYanJin commented on ui/src/types.js:
```diff
@@ -1,3 +1,13 @@
 export type Result<T> = { error: any } | T;
 export type APIResult<T> = Result<{ body: T }>;
+
+export type NodeAddress = {
+  internalIP: string,
+  name: string,
+};
+export type NodesAddresses = Array<NodeAddress>;
```
IMO, all the typing here in this file can be removed and type directly in various places.

ChengYanJin commented on ui/src/services/prometheus/fetchMetrics.js:
```diff
@@ -30,3 +33,139 @@ export function queryNodeFSSize(
+}
+
+export function queryNodeLoadMetrics(
+  instanceIP: $PropertyType<NodeAddress, 'internalIP'>,
```
```suggestion
  instanceIP: string,
```

ChengYanJin commented on ui/src/services/graphUtils.test.js:
```diff
@@ -0,0 +1,262 @@
+    },
+  },
+  {
+    status: 'fail',
```
```suggestion
    status: 'error',
```

ChengYanJin reviewed (commented):
My last change request ;) mainly about the typing. Thanks for addressing them.

theocerutti committed:
ui, dashboard: Add new dashboard layout and metrics card

Refs: #3088

theocerutti committed:
ui, tests: Add tests for graphutils and timespan hook

Refs: #3088

ChengYanJin reviewed (approved):
LGTM! Not the scope for this PR, but we need to decide the time format we want to use in graphs.

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.10`



The following branches will **NOT** be impacted:

* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`
* `development/2.9`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.10`



The following branches have **NOT** changed:

* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`
* `development/2.9`



Please check the status of the associated issue None.

Goodbye alexis-ld.
